jlm/llvm/opt/CommonNodeElimination.cpp

```diff
@@ -0,0 +1,1017 @@
+/*
+ * Copyright 2017 Nico Reißmann <nico.reissmann@gmail.com>
+ * Copyright 2025 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <jlm/llvm/ir/operators.hpp>
+#include <jlm/llvm/ir/RvsdgModule.hpp>
+#include <jlm/llvm/opt/CommonNodeElimination.hpp>
+#include <jlm/rvsdg/gamma.hpp>
+#include <jlm/rvsdg/MatchType.hpp>
+#include <jlm/rvsdg/theta.hpp>
+#include <jlm/rvsdg/traverser.hpp>
+#include <jlm/util/Statistics.hpp>
+#include <jlm/util/time.hpp>
+#include <map>
+
+namespace jlm::llvm
+{
+
+class CommonNodeElimination::Statistics final : public util::Statistics
+{
+  const char * MarkTimerLabel_ = "MarkTime";
+  const char * DivertTimerLabel_ = "DivertTime";
+
+public:
+  ~Statistics() override = default;
+
+  explicit Statistics(const util::FilePath & sourceFile)
+      : util::Statistics(Statistics::Id::CommonNodeElimination, sourceFile)
+  {}
+
+  void
+  startMarkStatistics(const rvsdg::Graph & graph) noexcept
+  {
+    AddMeasurement(Label::NumRvsdgInputsBefore, rvsdg::ninputs(&graph.GetRootRegion()));
+    AddTimer(MarkTimerLabel_).start();
+  }
+
+  void
+  endMarkStatistics() noexcept
+  {
+    GetTimer(MarkTimerLabel_).stop();
+  }
+
+  void
+  startDivertStatistics() noexcept
+  {
+    AddTimer(DivertTimerLabel_).start();
+  }
+
+  void
+  endDivertStatistics(const rvsdg::Graph & graph) noexcept
+  {
+    AddMeasurement(Label::NumRvsdgInputsAfter, rvsdg::ninputs(&graph.GetRootRegion()));
+    GetTimer(DivertTimerLabel_).stop();
+  }
+
+  static std::unique_ptr<Statistics>
+  Create(const util::FilePath & sourceFile)
+  {
+    return std::make_unique<Statistics>(sourceFile);
+  }
+};
+
+/**
+ * Class representing congruence sets over outputs.
+ * Different outputs can be grouped together into congruence sets,
+ * where all members of a set are considered identical.
+ *
+ * One output per set is the leader, and the rest are followers.
+ * During analysis of loops, outputs can be speculatively placed in the same congruence set.
+ * Later passes can then quickly verify if a outputs does not belong, by comparing against only
+ * the congruence set leader.
+ */
+class CongruenceSets
+{
+public:
+  /**
+   * Represents a set of T where all Ts in the set are considered identical.
+   * All the Ts must be in the same region.
+   * One T is the leader, and will eventually become the single source of truth.
+   */
+  struct CongruenceSet
+  {
+    // The set leader. Never changes.
+    // Once an output has become a leader, it will never be a follower.
+    // A leader always comes earlier than its followers in the traverser.
```

Comment:
I do not understand what "A leader always comes earlier than its followers in the traverser" means? Why is this piece of information here (whatever it means)?